diff --git b/AwsCryptographyPrimitives/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/primitives/AtomicPrimitives.java a/AwsCryptographyPrimitives/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/primitives/AtomicPrimitives.java
index 2652f9bb..bcaa2894 100644
--- b/AwsCryptographyPrimitives/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/primitives/AtomicPrimitives.java
+++ a/AwsCryptographyPrimitives/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/primitives/AtomicPrimitives.java
@@ -10,6 +10,9 @@ import java.lang.Byte;
 import java.lang.IllegalArgumentException;
 import java.nio.ByteBuffer;
 import java.util.Objects;
+// BEGIN MANUAL FIX
+import software.amazon.cryptography.primitives.internaldafny.AtomicPrimitivesClient;
+// END MANUAL FIX
 import software.amazon.cryptography.primitives.internaldafny.__default;
 import software.amazon.cryptography.primitives.internaldafny.types.Error;
 import software.amazon.cryptography.primitives.internaldafny.types.IAwsCryptographicPrimitivesClient;
@@ -44,8 +47,11 @@ public class AtomicPrimitives {
     CryptoConfig input = builder.CryptoConfig();
     software.amazon.cryptography.primitives.internaldafny.types.CryptoConfig dafnyValue =
       ToDafny.CryptoConfig(input);
-    Result<IAwsCryptographicPrimitivesClient, Error> result =
-      __default.AtomicPrimitives(dafnyValue);
+    // BEGIN MANUAL FIX
+    Result<AtomicPrimitivesClient, Error> result = __default.AtomicPrimitives(
+      dafnyValue
+    );
+    // END MANUAL FIX
     if (result.is_Failure()) {
       throw ToNative.Error(result.dtor_error());
     }
