diff --git b/TestModels/CodegenPatches/Model/SimpleCodegenpatchesTypes.dfy a/TestModels/CodegenPatches/Model/SimpleCodegenpatchesTypes.dfy
--- b/TestModels/CodegenPatches/Model/SimpleCodegenpatchesTypes.dfy
+++ a/TestModels/CodegenPatches/Model/SimpleCodegenpatchesTypes.dfy
@@ -158,8 +158,14 @@ abstract module AbstractSimpleCodegenpatchesService
       ensures GetStringEnsuresPublicly(input, output)
       ensures History.GetString == old(History.GetString) + [DafnyCallEvent(input, output)]
     {
-      output := Operations.GetString(config, input);
-      History.GetString := History.GetString + [DafnyCallEvent(input, output)];
+      // BEGIN MANUAL EDIT
+      var defaultedInput := input;
+      if input.value.None? {
+        defaultedInput := input.(value := Some("default"));
+      }
+      output := Operations.GetString(config, defaultedInput);
+      History.GetString := History.GetString + [DafnyCallEvent(defaultedInput, output)];
+      // END MANUAL EDIT
     }
 
   }
