diff --git a/AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/Keyrings/AwsKms/AwsKmsKeyring.dfy b/AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/Keyrings/AwsKms/AwsKmsKeyring.dfy
index b318ab09..3684bd23 100644
--- a/AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/Keyrings/AwsKms/AwsKmsKeyring.dfy
+++ b/AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/Keyrings/AwsKms/AwsKmsKeyring.dfy
@@ -386,7 +386,7 @@ module AwsKmsKeyring {
     //# OnDecrypt MUST take [decryption materials]
     //# (../structures.md#decryption-materials) and a list of [encrypted data
     //# keys](../structures.md#encrypted-data-key) as input.
-    method {:vcs_split_on_every_assert} OnDecrypt'(
+    method OnDecrypt'(
       input: Types.OnDecryptInput
     )
       returns (res: Result<Types.OnDecryptOutput, Types.Error>)
