diff --git b/AwsCryptographicMaterialProviders/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/keystore/KeyStore.java a/AwsCryptographicMaterialProviders/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/keystore/KeyStore.java
index 68bf0666..7efec786 100644
--- b/AwsCryptographicMaterialProviders/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/keystore/KeyStore.java
+++ a/AwsCryptographicMaterialProviders/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/keystore/KeyStore.java
@@ -6,6 +6,9 @@ package software.amazon.cryptography.keystore;
 import Wrappers_Compile.Result;
 import java.lang.IllegalArgumentException;
 import java.util.Objects;
+// BEGIN MANUAL FIX
+import software.amazon.cryptography.keystore.internaldafny.KeyStoreClient;
+// END MANUAL FIX
 import software.amazon.cryptography.keystore.internaldafny.__default;
 import software.amazon.cryptography.keystore.internaldafny.types.Error;
 import software.amazon.cryptography.keystore.internaldafny.types.IKeyStoreClient;
@@ -32,7 +35,9 @@ public class KeyStore {
     KeyStoreConfig input = builder.KeyStoreConfig();
     software.amazon.cryptography.keystore.internaldafny.types.KeyStoreConfig dafnyValue =
       ToDafny.KeyStoreConfig(input);
-    Result<IKeyStoreClient, Error> result = __default.KeyStore(dafnyValue);
+    // BEGIN MANUAL FIX
+    Result<KeyStoreClient, Error> result = __default.KeyStore(dafnyValue);
+    // END MANUAL FIX
     if (result.is_Failure()) {
       throw ToNative.Error(result.dtor_error());
     }
