diff --git b/TestVectorsAwsCryptographicMaterialProviders/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/materialproviderstestvectorkeys/KeyVectors.java a/TestVectorsAwsCryptographicMaterialProviders/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/materialproviderstestvectorkeys/KeyVectors.java
index 50632855..e9a071c5 100644
--- b/TestVectorsAwsCryptographicMaterialProviders/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/materialproviderstestvectorkeys/KeyVectors.java
+++ a/TestVectorsAwsCryptographicMaterialProviders/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/materialproviderstestvectorkeys/KeyVectors.java
@@ -10,6 +10,9 @@ import software.amazon.cryptography.materialproviders.CryptographicMaterialsMana
 import software.amazon.cryptography.materialproviders.ICryptographicMaterialsManager;
 import software.amazon.cryptography.materialproviders.IKeyring;
 import software.amazon.cryptography.materialproviders.Keyring;
+// BEGIN MANUAL FIX
+import software.amazon.cryptography.materialproviderstestvectorkeys.internaldafny.KeyVectorsClient;
+// END MANUAL FIX
 import software.amazon.cryptography.materialproviderstestvectorkeys.internaldafny.__default;
 import software.amazon.cryptography.materialproviderstestvectorkeys.internaldafny.types.Error;
 import software.amazon.cryptography.materialproviderstestvectorkeys.internaldafny.types.IKeyVectorsClient;
@@ -29,7 +32,9 @@ public class KeyVectors {
     KeyVectorsConfig input = builder.KeyVectorsConfig();
     software.amazon.cryptography.materialproviderstestvectorkeys.internaldafny.types.KeyVectorsConfig dafnyValue =
       ToDafny.KeyVectorsConfig(input);
-    Result<IKeyVectorsClient, Error> result = __default.KeyVectors(dafnyValue);
+    // BEGIN MANUAL FIX
+    Result<KeyVectorsClient, Error> result = __default.KeyVectors(dafnyValue);
+    // END MANUAL FIX
     if (result.is_Failure()) {
       throw ToNative.Error(result.dtor_error());
     }
