diff --git b/AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/Model/AwsCryptographyKeyStoreTypes.dfy a/AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/Model/AwsCryptographyKeyStoreTypes.dfy
index 26600cfa..a3d5c6c9 100644
--- b/AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/Model/AwsCryptographyKeyStoreTypes.dfy
+++ a/AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/Model/AwsCryptographyKeyStoreTypes.dfy
@@ -285,7 +285,9 @@ abstract module AbstractAwsCryptographyKeyStoreService
   import Operations : AbstractAwsCryptographyKeyStoreOperations
   function method DefaultKeyStoreConfig(): KeyStoreConfig
   method KeyStore(config: KeyStoreConfig := DefaultKeyStoreConfig())
-    returns (res: Result<IKeyStoreClient, Error>)
+    // BEGIN MANUAL FIX
+    returns (res: Result<KeyStoreClient, Error>)
+    // END MANUAL FIX
     requires config.ddbClient.Some? ==>
                config.ddbClient.value.ValidState()
     requires config.kmsClient.Some? ==>
@@ -314,10 +316,14 @@ abstract module AbstractAwsCryptographyKeyStoreService
               config.kmsClient.value.ValidState()
 
   // Helper function for the benefit of native code to create a Success(client) without referring to Dafny internals
-  function method CreateSuccessOfClient(client: IKeyStoreClient): Result<IKeyStoreClient, Error> {
+  // BEGIN MANUAL FIX
+  function method CreateSuccessOfClient(client: KeyStoreClient): Result<KeyStoreClient, Error> {
+    // END MANUAL FIX
     Success(client)
   } // Helper function for the benefit of native code to create a Failure(error) without referring to Dafny internals
-  function method CreateFailureOfError(error: Error): Result<IKeyStoreClient, Error> {
+  // BEGIN MANUAL FIX
+  function method CreateFailureOfError(error: Error): Result<KeyStoreClient, Error> {
+    // END MANUAL FIX
     Failure(error)
   }
   class KeyStoreClient extends IKeyStoreClient
